### PR TITLE
#2451 - Enable typescript reference on backend errors

### PR DESCRIPTION
**Pull request by andrepestana-aot**
- Added flag `--enable-source-maps` to node apps;
- Added `devtool: "source-map"` to webpack.config.js. Followed this comment about monorepo: https://github.com/nestjs/nest/issues/4644#issuecomment-616992927
- Fixed some errors being suppressed found;
- Created a ticket to enforce the error throwing to be an instance of `Error` in TS: https://app.zenhub.com/workspaces/student-information-management-system-5fce9df5aa1b45000e937014/issues/gh/bcgov/sims/2773
- Created a ticket to update NestJS in order to allow passing `cause` to HTTP Exceptions: https://app.zenhub.com/workspaces/student-information-management-system-5fce9df5aa1b45000e937014/issues/gh/bcgov/sims/2774
- Expected logs in containers now:
image
image

### Notes:
- If an error is thrown as string, the line number is not shown:
image
image
That may require to change TS rules [in a new ticket](https://app.zenhub.com/workspaces/student-information-management-system-5fce9df5aa1b45000e937014/issues/gh/bcgov/sims/2773) to enforce thrown errors to be an instance of `Error`: https://typescript-eslint.io/rules/no-throw-literal/
- The current @nestjs/common version used doesn't have support for passing `cause` in the HTTP Exception:
    - Current used version: https://github.com/nestjs/nest/blob/v9.0.3/packages/common/exceptions/http.exception.ts
    - Latest version: https://github.com/nestjs/nest/blob/master/packages/common/exceptions/http.exception.ts
    It requires to update NestJS [in another ticket](https://app.zenhub.com/workspaces/student-information-management-system-5fce9df5aa1b45000e937014/issues/gh/bcgov/sims/2774) to allow passing `cause` to HTTP exceptions.

### CODE REVIEW

File: sources/packages/backend/package.json [hunk 1]
```diff
@@ -13,14 +13,14 @@
...
-    "start:prod:load-test-gateway": "node dist/apps/load-test-gateway/main.js",
-    "docker:start:load-test-gateway": "node dist/apps/load-test-gateway/main.js",
+    "start:prod:api": "node --enable-source-maps dist/apps/api/main.js",
+    "docker:start:api": "node --enable-source-maps dist/apps/api/main.js",
+    "start:prod:workers": "node --enable-source-maps dist/apps/workers/main.js",
+    "docker:start:workers": "node --enable-source-maps dist/apps/workers/main.js",
+    "start:prod:queue-consumers": "node --enable-source-maps dist/apps/queue-consumers/main.js",
+    "docker:start:queue-consumers": "node --enable-source-maps dist/apps/queue-consumers/main.js",
+    "start:prod:load-test-gateway": "node --enable-source-maps dist/apps/load-test-gateway/main.js",
+    "docker:start:load-test-gateway": "node --enable-source-maps dist/apps/load-test-gateway/main.js",
```
Comment (sh16011993) [inline sources/packages/backend/package.json:23]: FMU, don't we need to add `source-map-support` as a dependency?

Comment (andrepestana-aot) [reply sources/packages/backend/package.json:23]: NodeJS has already support for that from version 12.12: https://nodejs.org/en/blog/release/v12.12.0/

**Comment by guru-aot** [comment]
This means the cause will not be shown until we upgrade the version and after that do we need to have any change other than the required one?

File: sources/packages/backend/webpack.config.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
 const CopyPlugin = require("copy-webpack-plugin");
 module.exports = {
+  devtool: "source-map",
```
Comment (guru-aot) [inline sources/packages/backend/webpack.config.js:3]: 👍

**Review by andrewsignori-aot — APPROVED** [review]
Nice work, looks good 👍 
Can we add an entry in the wiki about ensuring that an error throw will always contain the cause. This should also be part of the PR reviews.

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrepestana-aot

**Review by lewischen-aot — APPROVED** [review]
LGTM 👍 Nice work!

**Review by dheepak-aot — APPROVED** [review]
Looks Good 👍

**Review by sh16011993 — APPROVED** [review]
Nice work @andrepestana-aot 👍

**Comment by andrepestana-aot** [comment]
> This means the cause will not be shown until we upgrade the version and after that do we need to have any change other than the required one?

@guru-aot The NestJS version is just to allow us to pass the cause to HTTP Exceptions when it is appropriate. Then, after the update, we can check if it would be good to add a cause to them. All the other cases are already working when we pass cause to the `Error` instantiation like in https://github.com/bcgov/SIMS/blob/main/sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-part-time-integration/e-cert-files/e-cert-file-record.ts#L225

**Comment by andrepestana-aot** [comment]
> Nice work, looks good 👍 Can we add an entry in the wiki about ensuring that an error throw will always contain the cause. This should also be part of the PR reviews.

Sure. Will do that.

**Pull request merged**